Lcg/1923

lamont-granquist opened the pull request:
validatorless bootstrapping and chef-vault integration

lamont-granquist commented:
@chef/client-core @chef/client-engineers 

also, @afiune i finally got this done.

lamont-granquist commented:
also @moserke, @dougireton and @jf647 you should be interested in this

stevendanna commented on lib/chef/knife/bootstrap/client_builder.rb:
```diff
@@ -0,0 +1,190 @@
+        #
+        # We prompt the user about what to do and will fail hard if we do not get confirmation to
+        # delete any prior node/client objects.
+        def sanity_check
```
I don't think it should block this for now, but perhaps our "model" classes should just provide this functionality via class methods.  `Chef::Client.exists?(name)`, `Chef::Client.delete(name)`, `Chef::Client.delete_f(name)` (or something for "delete but I don't care if it isn't there").

lamont-granquist replied:
Yep, we've got that in a few places, and I agree.

btm commented:
It's confusing to me that these classes take a chef_server_rest argument that gets set to the instance variable of the same name, but there's also a method named chef_server_rest that produces something to be set to that instance variable. I presume in initialize that chef_server_rest is the argument and not the method, but is it necessary for these names to overlap?

lamont-granquist commented:
Hmm, I guess I'm unclear on what's confusing.  The objects right now have a hard dependency on Chef::REST and on a particular way to create a new rest object there and its Chef::Config values.  By allowing the caller to inject that object instead of having the class create it there's a looser dependency so that you can inject a different Chef::REST object (like I'm using to inject a rest object with the authz/authn of the client that is being created instead of the user running knife), or you could even inject some other kind of object that duck-types to a Chef::REST interface.

It makes sense to me because when the caller supplies a chef_server_rest parameter to the constructor that gets assigned to the chef_server_rest ivar and is retrieved later by the chef_server_rest accessor.  All that you do via injecting it into the constructor is avoid the lazy `||=` default construction of the object.

lamont-granquist commented:
@chef/client-core @chef/client-engineers re-paging for review, would like to hit merge before monday.

jf647 commented:
Looks like a nice addition to knife bootstrap.  We still have to solve the initial registration flow more gracefully for things like autoscaling and cloudformation, but that's on the chef-vault 3.x roadmap.

Would you be open to making the options

    knife[:bootstrap_vault_whatever]

rather than

    knife[:vault_whatever]

?  I'm concerned that we have options in a flat namespace, some of which serve as stand-ins for switches to 'knife vault' and some of which server as stand-ins for switches to 'knife bootstrap'.

Is there anything in the public-facing chef-vault API that would make this easier for you?  I note that if the search fails we just emit to stdout, but given that you're waiting for SOLR to index the new node and client that should be OK.  We could raise errors instead in ChefVault::Item and have the existing knife commands catch those and emit similar error strings for compatibility.

lamont-granquist commented:
Yeah I can namespace it all easy, very happy to do that and not re-invent the knife[:image] collision problem...

If you can make chef-vault accept the generated client and not require search that'd be best.  This code creates a client then saves it then retrieves it via solr and that whole round trip seems pointless.  I think part of that is that the chef-client API itself may not support what I want to do, but if you could look at what is needed from the chef-vault side maybe you can see how to fix it?

tyler-ball commented on lib/chef/cookbook_version.rb:
```diff
@@ -532,7 +535,7 @@ def self.chef_server_rest
 
     def chef_server_rest
-      self.class.chef_server_rest
+      @chef_server_rest ||= Chef::REST.new(Chef::Config[:chef_server_url])
```
This could be DRYed up with @chef_server_rest ||= self.class.chef_server_rest

lamont-granquist replied:
Except those class methods with global state are terrible, I'd rather go the other direction and make all the class methods call new to create an instance to do their work.

lamont-granquist replied:
Actually I see its not using any global state in the class, its still pretty ugly because of the tight coupling, I'm just not sure what to do about it.

tyler-ball commented on lib/chef/knife/core/bootstrap_context.rb:
```diff
@@ -42,7 +44,11 @@ def bootstrap_environment
         def validation_key
-          IO.read(File.expand_path(@chef_config[:validation_key]))
+          if File.exist?(File.expand_path(@chef_config[:validation_key]))
+            IO.read(File.expand_path(@chef_config[:validation_key]))
```
What do people think about an info level log here to say that bootstrap is using the local validation_key?  I ran `knife configure` and I think it left my `validation_key` in my `knife.rb`.  If I use the new register-node-locally logic, I probably don't want that accidently sent to bootstrapped nodes.

tyler-ball commented:
One thing I missed - when are users supposed to call the new `Chef::DataBag.new(server_url)` initialization method?  I see that parameter added to all the models but I don't know why.

Other than that and some minor comments, it LGTM.  I like the idea of registering a node before bootstrapping it!

lamont-granquist commented:
I added that parameter for consistency.  I needed to DI the REST class with the client object into the Chef::Node so I needed it there, and I added it a few other places to reduce the tight coupling and allow for other DI uses.  There's a YAGNI argument against it, but I'd rather try to keep the API evolving consistently.

lamont-granquist commented:
okay rebased on master

adamedx commented:
@lamont, @btm, do we need to change knife-windows to support this?

adamedx commented:
@btm, we should test knife-windows against this version of Chef -- with Chef Client 12.0.0, we broke knife-windows in multiple ways, we should ensure existing functionality is not regressed.

lamont-granquist commented:
I certainly tried to be careful about that, which is why default_bootstrap_template and server_name are now documented APIs.

lamont-granquist commented:
Can I get a few actual :+1:'s before monday?

jf647 commented:
@lamont-granquist, I've pushed a [branch](https://github.com/Nordstrom/chef-vault/tree/chef_lcg_1923) that allows ChefVault::Item#clients to accept a Chef::ApiClient.  Take a look and see if it will work for you.

I can get it on master pretty quick, but the next gem release wasn't going to be for a bit (working through an issue backlog).  Do you need it on rubygems to use?  I gather from Jay's request to tag a release that was lacking one that the chef-dk omnibus builds pull from GH directly.

danielsdeleo commented:
:+1: 

An nitpick: I dislike calling the API object 'chef_server_rest' (REST is a concept, not a thing you can have several of), which is why I use 'api_service' or something like that everywhere. But it's consistent with the class name and what we do elsewhere so I can live with it.

lamont-granquist commented:
Yeah I wasn't super wild about the name, but its what we've already named the thing.  Since its a named parameter we can change it later and deprecate the old one in favor of the new one.

lamont-granquist commented:
@jf647 i'm leaning towards shipping this like it is right now, and leaving that improvement for later. 

I'm trying to get this merged for 12.1.0 and we freeze for that on monday.

That is, however, exactly the patch that I think I need...

tyler-ball commented:
:+1: from me

afiune commented:
Red it and I want to say => :+1: but I just need to test it! (tomorrow!)
